### PR TITLE
[Flow] Define Flow types for style and stylesheet objects, more checking

### PR DESCRIPTION
**Pull request by ide**
- Define a Styles type for the object that gets passed into StyleSheet.create
- Define a StyleSheet type that is returned from StyleSheet.create
- Clean up the type declarations in StyleSheet.create

Test Plan: `flow`

### CODE REVIEW

**Comment by facebook-github-bot-8** [comment]
By analyzing the blame information on this pull request, we identified @sahrens and @prometheansacrifice to be potential reviewers.

**Comment by ide** [comment]
cc @vjeux @brentvatne

**Comment by vjeux** [comment]
@yungsters can you take a look?

**Comment by vjeux** [comment]
Oh this is actually not adding any new strictness, I can just ship it, thanks

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by vjeux** [comment]
Can you rebase?

**Comment by facebook-github-bot-7** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by satya164** [comment]
@facebook-github-bot shipit

File: Libraries/StyleSheet/StyleSheet.js [hunk 1]
```diff
@@ -17,6 +17,9 @@ var StyleSheetValidation = require('StyleSheetValidation');
 
 var flatten = require('flattenStyle');
 
+export type Styles = {[key: string]: Object};
```
Comment (satya164) [inline Libraries/StyleSheet/StyleSheet.js]: Could be `{[key: string]: {[key: string]: any} }` right?

**Comment by facebook-github-bot-5** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/663970937083484/int_phab) to review internal test results.

**Comment by vjeux** [comment]
It finds some more internal bad code. Someone needs to fix them internally first.

File: Libraries/StyleSheet/StyleSheet.js (hunk 1 above)
Comment (ide) [inline Libraries/StyleSheet/StyleSheet.js]: Isn't `{[key: string]: any}` the same as Object?

Comment (satya164) [inline Libraries/StyleSheet/StyleSheet.js]: @ide Not exactly sure how Flow works here. Do arrays also count as objects? If not then they are same :)

Comment (ide) [inline Libraries/StyleSheet/StyleSheet.js]: Yep:

screenshot 2016-07-20 12 07 32

Comment (satya164) [inline Libraries/StyleSheet/StyleSheet.js]: @ide Cool then

**Comment by facebook-github-bot-4** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by javache** [comment]
For some reason, internally this is failing with

```
react-native-github/Libraries/StyleSheet/StyleSheet.js:20
 20: export type Styles = {[key: string]: Object};
                                          ^^^^^^ object type. This type is incompatible with
 21: export type StyleSheet<S: Styles> = {[key: $Keys<S>]: number};
                                                           ^^^^^^ number

react-native-github/Libraries/StyleSheet/StyleSheet.js:21
 21: export type StyleSheet<S: Styles> = {[key: $Keys<S>]: number};
                                                           ^^^^^^ number. This type is incompatible with
 20: export type Styles = {[key: string]: Object};
                                          ^^^^^^ object type


Found 2 errors
```

**Comment by ide** [comment]
That error doesn't make sense to me, reading the code and running `flow` (v0.30) myself. Perhaps a different version of Flow? Will try again.

@facebook-github-bot shipit

**Comment by facebook-github-bot-2** [comment]
I will not rubber stamp and land your change for you @ide!  I can import it for you and you can get your change reviewed by someone though :)

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-7** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3587964) to review internal test results.

**Comment by facebook-github-bot-5** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by gabelevi** [comment]
It's possible that improving the types like this exposes mistakes in user code. If you need help I can assist in the merge on Monday

**Comment by gabelevi** [comment]
Ok, fixed this up and landing it!

The flow error wasn't occurring within the react-native code, but rather in some code that uses react-native. Specifically, it was a jest mock that mocked `StyleSheet.create()` with an incorrect type. The error message was unfortunate in that it didn't point out the location of the mistake until you ran `flow check --traces 10`

**Comment by ide** [comment]
@gabelevi thanks so much for looking into the issue and merging this =)

**Pull request closed without merging**